luarules/gadgets/cmd_nano_range_check.lua

```diff
@@ -0,0 +1,156 @@
+	local cmdCache = {}
+
+	for nanoID, targets in pairs(trackingTable) do
+		if nanoID % updateInterval == frame % updateInterval then
+			if next(targets) == nil then
+				trackingTable[nanoID] = nil
+			end
+
+			local nanoDefID = spGetUnitDefID(nanoID)
+			local maxDistance = constructionTurretsDefs[nanoDefID].maxBuildDistance
+
+			for targetID, commandData in pairs(targets) do
+				if spValidUnitID(targetID) then
+					if commandData.cmdTag == -1 then
+						if tagUpdates >= tagUpdateFrameMax then break end
+
+						if not cmdCache[nanoID] then
+							cmdCache[nanoID] = spGetUnitCommands(nanoID, -1)
+							tagUpdates = tagUpdates +1
+						end
+
+						for _, cmd in ipairs(cmdCache[nanoID]) do
+							trackingTable[nanoID][cmd.params[1]].cmdTag = cmd.tag
+						end
+					end
+
+					local distance = spGetUnitSeparation(nanoID, targetID, constructionTurretsDefs[nanoDefID].buildRange3D, false)
+
+					if distance < maxDistance then
+						if commandData.status == statuses.outOfRange then
+							commandData.status = statuses.inRange
+						end
+					else
+						-- Do not blindly remove orders on out-of-range mobile units, as the player may have pre-emptively issued an order before the unit was in range
+						-- Instead, only remove the order once the unit has come into range then left again
+						if commandData.status == statuses.inRange then
+							spGiveOrderToUnit(nanoID, CMD.REMOVE, commandData.cmdTag, 0)
+							targets[targetID] = nil
+						end
+					end
+				else
+					targets[targetID] = nil
+				end
+			end
+		end
+	end
+end
+
+function gadget:AllowCommand(unitID, unitDefID, teamID, cmdID,	cmdParams, cmdOptions, cmdTag, synced, fromLua)
+	local constructionTurretDef = constructionTurretsDefs[unitDefID]
+	if not constructionTurretDef then return true end
+
+	if cmdID == CMD_STOP then
+		trackingTable[unitID] = nil
+	end
+
+	-- Rather than trying to handle area commands, let them fallthrough, then catch the resulting list of single-target commands
+	if #cmdParams ~= 1 then return true end
+
+	local distance = math.huge
+	local targetId = cmdParams[1]
+
+	if targetId < gMaxUnits then
+		if not spValidUnitID(targetId) then return end
+		local defID = spGetUnitDefID(targetId)
+
+		if mobileUnits[defID] then
+			trackingTable[unitID] = trackingTable[unitID] or {}
+			trackingTable[unitID][cmdParams[1]] = {cmdTag = -1, status = statuses.outOfRange}
+			return true
+		end
+
+		distance = spGetUnitSeparation(unitID, targetId, constructionTurretDef.buildRange3D, false)
+	else
+		targetId = targetId - gMaxUnits
+		if not spValidFeatureID(targetId) then return end
+		distance = spGetUnitFeatureSeparation(unitID, targetId, false)
+	end
+
+	if distance > constructionTurretDef.maxBuildDistance then
+		return false
+	end
+
+	return true
+end
+
+function gadget:UnitDestroyed(unitID)
+	if trackingTable[unitID] then
+		trackingTable[unitID] = nil
+	end
+
+	for nanoID, _ in pairs(trackingTable) do
+		trackingTable[nanoID][unitID] = nil
+	end
```

Comment:
The check order could be swapped (maybe in the future somebody adds self-repairing and then a nano is its own target). The code could also be simplified a bit:
```suggestion
	for _, targets in pairs(trackingTable) do
		targets[unitID] = nil
	end
	trackingTable[unitID] = nil
```

Reply:
Swapping them does nothing, in fact it's marginally worse. If the destroyed unit is a nano, remove all the targets that nano was tracking, since now it's gone. Then, go through all the nanos and if they were tracking the destroyed unit, remove that single targeting entry.

Swapping the order means iterating through the targets of the destroyed nano, then getting rid of all of them anyway.

Reply:
Right, there isn't actually any extra data to clean up.